File: cc/leetcode/string/828_count_unique.cc
class Solution {
public:
    int uniqueLetterString(string s) {
        vector<vector<int>> memo(26, vector<int>(0));
        int ns = s.size();
        for (int i = 0; i < ns; ++i) {
            int idx = s[i] - 'A';
            memo[idx].push_back(i);
        }
        long res = 0;
        long MOD = pow(10, 9) + 7;
        int left, right;
        
        for (int i=0; i < 26; ++i) {
            int n = memo[i].size();
            if (n == 0)
                continue;
            for (int j = 0; j < n; ++j) {
                if (j == 0)
                    left = 0;
                else
                    left = memo[i][j-1]+1;
                if (j == n-1)
                    right = ns -1;
                else
                    right = memo[i][j+1]-1;
                res = (res + (memo[i][j]-left+1) * (right-memo[i][j]+1)) % MOD;
            }
        }
        return res;
    }
};

File: python/leetcode/string/828_count_unique.py
"""
828. Count Unique Characters of All Substrings of a Given String (Hard)

"""

class Solution(object):
    def uniqueLetterString(self, s):
        """
        :type s: str
        :rtype: int
        """
        memo = []
        for i in range(26):
            memo.append([])
            
        for i, c in enumerate(s):
            idx = ord(c) - ord('A')
            memo[idx].append(i)
            
        size = len(s)
        MOD = 10 ** 9 + 7
        
        res = 0
        for i in range(26):
            n = len(memo[i])
            if n == 0:
                continue
            for j in range(n):
                if j > 0:
                    left = memo[i][j-1] + 1
                else:
                    left = 0
                if j < n - 1:
                    right = memo[i][j+1] - 1
                else:
                    right = size - 1
                res = (res + (memo[i][j] - left + 1) * (right - memo[i][j] + 1)) % MOD
        return res
        
